Columns Block: Add custom gutter widths

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/10730

## Description
Adds the ability to set custom gutter widths for a columns block via a new attribute and inline style. 

**No new UI control will be added until the new sidebar controls and component system is available.**

## How has this been tested?
Manually.

#### Testing Instructions

1. Create a new post adding a columns block 
  (it can help make the gutters more obvious if you add inner group blocks with a background color)
2. Checkout this PR and edit the post
3. Select the columns block and confirm no gutter size controls present in sidebar
4. Switch to code editor view and add a `gutterSize` attribute to the columns block along with a matching inline style. e.g. `--columns-block-gutter-size: 0px;` or simply copy and paste in the block code below:

```
<!-- wp:columns {"gutterSize":2,"backgroundColor":"primary"} -->
<div class="wp-block-columns has-primary-background-color has-background" style="--columns-block-gutter-size:2px"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:group {"backgroundColor":"accent"} -->
<div class="wp-block-group has-accent-background-color has-background"><div class="wp-block-group__inner-container"></div></div>
<!-- /wp:group --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:group {"backgroundColor":"accent"} -->
<div class="wp-block-group has-accent-background-color has-background"><div class="wp-block-group__inner-container"></div></div>
<!-- /wp:group --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

5. Confirm the column gutters update appropriately within the editor and on the frontend after saving
6. Add nested columns and add differing gutter sizes or paste in code from this [example block content](https://gist.github.com/aaronrobertshaw/861402fe1a809510c3ac0679941eb99e)
7. Confirm the nested columns have different gutter widths

## Screenshots

<img width="1465" alt="Screen Shot 2020-12-23 at 7 40 32 pm" src="https://user-images.githubusercontent.com/60436221/102983220-79087200-4557-11eb-8781-046a2251a74b.png">

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

aaronrobertshaw committed:
Add custom gutter widths to columns

glendaviesnz commented on packages/block-library/src/columns/edit.js:
```diff
@@ -69,6 +75,10 @@ function ColumnsEditContainer( {
 	const blockProps = useBlockProps( {
 		className: classes,
+		style: {
+			'--gutter-size':
```
I wonder if we should make the namespacing clearer here, eg. --columns-block-gutter-size, just to make it clearer when debugging in dev tools, etc. what this var is related to ?

aaronrobertshaw replied:
Given the CSS var was not set on the `:root` selector and only via the `.wp-block-columns` scope I felt it was clear enough where this gutter size was coming from and its purpose. 

It doesn't hurt to improve clarity of course so I've made the suggested change.  Thanks 🙂

glendaviesnz commented:
This tested well for me.  My only question is with nested column blocks where parent gutter setting flows through to any child columns block that hasn't explicitly specified a gutter. Once a child explicitly sets the gutter size it does override the parent setting. This seems reasonable to me, but just wondering if it is the expected behaviour, or if it will confuse users at all when next columns?

aaronrobertshaw committed:
Prevent undesired inheritance of gutter size

aaronrobertshaw committed:
Rename CSS var for improve clarity of source

aaronrobertshaw commented:
> This tested well for me. My only question is with nested column blocks where parent gutter setting flows through to any child columns block that hasn't explicitly specified a gutter.

That is the nature of CSS, I don't think our users would have the same perspective though. I can see benefits either way. The UI doesn't really inform the user that the setting would cascade. In this case it might be better to require the gutter sizes to be explicitly set on the nested columns.

I've made that change, let me know whether you think it's an improvement.

glendaviesnz commented:
To me this is now works more how you would expect, ie. changes a `columns` gutter only affects that columns block and not all the children ... but will be interesting to see what others think.

stacimc commented:
+1 for the setting not automatically applying to nested columns; I also think that feels more natural/how I would expect those settings to work as a user. This tested well for me.

aaronrobertshaw committed:
Remove UI controls

Until the new sidebar controls and system lands, it is preferred to avoid adding any new UI controls to the sidebar.

apeatling commented:
> Switch to code editor view and add a gutterSize attribute to the columns block along with a matching inline style. e.g. --columns-block-gutter-size: 0px;

What exactly should this look like in the code view? It's not clear.

apeatling reviewed (commented):
> Switch to code editor view and add a gutterSize attribute to the columns block along with a matching inline style. e.g. --columns-block-gutter-size: 0px;

What exactly should this look like in the code view? It's not clear.

aaronrobertshaw commented:
> What exactly should this look like in the code view? It's not clear.

Sorry about that. I'll update the test instructions with an explicit code example. The linked gist from step 6 also contains example block code with the attribute and inline style set.

```
<!-- wp:columns {"gutterSize":2,"align":"wide"} -->
<div class="wp-block-columns alignwide" style="--columns-block-gutter-size:2px"><!-- wp:column -->
```

apeatling reviewed (approved):
Thanks for the code snippet. This worked well for me. 👍

aaronrobertshaw commented:
A [new iteration of the block supports mechanism](https://github.com/WordPress/gutenberg/issues/28913) is under way (https://github.com/WordPress/gutenberg/issues/28913). 

Those efforts aim to provide a means to target which inner elements get the block support style applied to them. With that functionality, the columns block could opt into a new spacing > margins block support feature and apply those styles to the inner columns.

The current approach in this PR is relatively clean but does not offer the ability to set the gutter sizes via theme.json.  I'll revisit this once the new iteration of block supports lands.

aaronrobertshaw commented:
Closing this PR. In some environments kses will strip out CSS custom properties from the saved post content. This means when the block is loaded again in the editor is will be deemed invalid.

The new approach to applying block support styles to inner blocks is likely to be the best path forward for this. That is, once margins are added to the current spacing block support.
